internal/pkg/cosign/payload/signer.go

```diff
@@ -38,40 +38,57 @@ var _ cosign.Signer = (*payloadSigner)(nil)
 
 // Sign implements `Signer`
 func (ps *payloadSigner) Sign(ctx context.Context, payload io.Reader) (oci.Signature, crypto.PublicKey, error) {
-	payloadBytes, err := io.ReadAll(payload)
+	payloadBytes, sig, pk, err := ps.signPayload(ctx, payload)
 	if err != nil {
 		return nil, nil, err
 	}
-	sOpts := []signature.SignOption{signatureoptions.WithContext(ctx)}
-	sOpts = append(sOpts, ps.payloadSignerOpts...)
-	sig, err := ps.payloadSigner.SignMessage(bytes.NewReader(payloadBytes), sOpts...)
+
+	b64sig := base64.StdEncoding.EncodeToString(sig)
+	ociSig, err := static.NewSignature(payloadBytes, b64sig)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	pkOpts := []signature.PublicKeyOption{signatureoptions.WithContext(ctx)}
-	pkOpts = append(pkOpts, ps.publicKeyProviderOpts...)
-	pk, err := ps.payloadSigner.PublicKey(pkOpts...)
+	return ociSig, pk, nil
+}
+
+func (ps *payloadSigner) signPayload(ctx context.Context, payload io.Reader) (payloadBytes, sig []byte, pk crypto.PublicKey, err error) {
+	payloadBytes, err = io.ReadAll(payload)
```

Comment:
There are a lot of paths where we `io.ReadAll` and `bytes.NewBuffer(someBytes)` around this signature.

I'm like 95% certain that if you kill this return value and forced the caller to read it themselves and pass a `bytes.NewReader` that the overall change would be a bunch simpler and clearer.

4 return values is a lot and this is sorta starting to have a code smell of 1 function doing too many things (as evidenced by its signature).  Already a better name would be `readAndSignPayload`, but I think we should drop the return value, which kills the `readAnd` bit.

Reply:
Receiving `io.Reader`s for signatures and payloads is an unexpectedly popular request, but I'd be OK supporting an interface which takes byte arrays

Reply:
I was actually suggesting we keep the `io.Reader`, but shuffling things so there are fewer layers of `bytes.NewBuffer` by reorienting expectations of who reads what.  That said, the only meaningful change here is in private stuff, so we can freely change things later :)